### ISSUE TITLE
Implement new Concept Exercise: iterators

### ISSUE THREAD

**Issue by workingjubilee** [exercism/rust#1065]
This issue describes how to implement the iterator concept exercise for the Rust track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

Iterators are one of Rust's more versatile linguistic abstractions, so it is not possible to cover in a single exercise. This exercise has only the goal of covering the foundations of Iterator so that advanced usage can be covered in detail.

## Learning objectives

Student should be able to...
- recognize some cases where they can use `.iter()`
- use `.next()` to advance the iterator manually
- use one `Iterator` method with a closure, such as `fold`, `map`, or `filter`
- use `fold` or `collect` to produce a single value
- recognize basic implications of iterator laziness (so a fold or collect is required)

Student may be able to...
- recognize Iterator is as or more performant than `for` loops

## Out of scope

- zip, chain, flatten, and other multi-iterator methods
- skip, step_by, take_while, and other selective iteration methods

## Concepts

- Iterator

## Prerequisites

- methods and traits
- in particular method chaining
- the collection in use (probably `vec` or `String`)
- some awareness of what an expression actually does
- some awareness of what a closure is like and/or higher-order functions
- if using filter: booleans and predicates

### Hints

- [std::iter::Iterator](https://doc.rust-lang.org/std/iter/trait.Iterator.html)
- Rust by Example: [Iterators](https://doc.rust-lang.org/stable/rust-by-example/trait/iter.html)
- The Rust Programming Language: [Processing Items with a Series of Iterators](https://doc.rust-lang.org/book/ch13-02-iterators.html)

### After

- TRPL on [iterator performance](https://doc.rust-lang.org/book/ch13-04-performance.html)
- Hoverbear going [in-depth on iterators](https://hoverbear.org/blog/a-journey-into-iterators/)

Likely better for later exercises, but relevant:
- The [itertools](https://docs.rs/itertools) crate for Even More Fun
- [Llogiq on performance](https://llogiq.github.io/2017/06/01/perf-pitfalls.html)

## Representer

This should require no special changes to the representer.

## Analyzer

This should require no special changes to the analyzer.

## Implementing

See our [implementation guide](https://github.com/exercism/v3/blob/master/languages/rust/howto/implementing-a-concept-exercise.md#how-to-implement-a-rust-concept-exercise) for guidance.
. You may want to make use of `boil_exercise.sh` in the `concept` directory

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

**Comment by nahuakang** [comment]
@coriolinus I wouldn't mind working on this next (both the concept and concept exercise). But it seems like we should have covered the `trait` concept before building out `iterator`? Do you have an opinion what would be the concepts/concept exercises that should be ready before working on `iterator`? I'd be happy to work on those first and eventually cover this issue.

**Comment by coriolinus** [comment]
It looks to me as though this concept is about using iterators, not
defining them. I'm not sure how useful it would be to require knowledge of
traits first. My instinct is that it should be OK to have this concept
depend solely on the methods concept, and then have students implement a
generator (via the Iterator trait) as part of the traits concept, which
naturally would then come after this one.

We have to mention in the concept documentation that it is a trait, but I
think it's OK to defer further explanation for later.

**Comment by nahuakang** [comment]
Happy to claim this if you assign it to me. Shall I also create an issue for the `trait` concept exercise?

**Comment by coriolinus** [comment]
Sounds good! Please do create a `traits` concept issue.

**Comment by senekor** [comment]
After several failed attempts to create a high-quality syllabus, I believe any future attempts will have to do their own design work from scratch.
